pkg/cli/predict.go

```diff
@@ -277,152 +395,264 @@ func predictIndividualInputs(predictor predict.Predictor, inputFlags []string, o
 		}
 
-		if outputPath == "" {
-			console.Output(s)
-		} else {
-			err := writeOutput(outputPath, []byte(s))
+		if writeOutputToDisk {
+			path, err := writeFile([]byte(s), outputPath)
 			if err != nil {
 				return fmt.Errorf("Failed to write output: %w", err)
 			}
+			console.Infof("Written output to: %s", path)
+		} else {
+			console.Output(s)
 		}
 
 		return nil
 	default:
-		// Treat everything else as JSON -- ints, floats, bools will all convert correctly.
-		rawJSON, err := json.Marshal(prediction.Output)
+		// Treat everything else as JSON -- ints, floats, bools will all be presented
+		// as raw values. Lists and objects will be pretty printed JSON.
+		output, err := prettyJSONMarshal(prediction.Output)
 		if err != nil {
-			return fmt.Errorf("Failed to encode prediction output as JSON: %w", err)
-		}
-		var indentedJSON bytes.Buffer
-		if err := json.Indent(&indentedJSON, rawJSON, "", "  "); err != nil {
 			return err
 		}
 
-		if outputPath == "" {
-			console.Output(indentedJSON.String())
-		} else {
-			err := writeOutput(outputPath, indentedJSON.Bytes())
+		// No special handling for needsJSON here.
+		if writeOutputToDisk {
+			path, err := writeFile(output, outputPath)
 			if err != nil {
 				return fmt.Errorf("Failed to write output: %w", err)
 			}
+			console.Infof("Written output to: %s", path)
+		} else {
+			console.Output(string(output))
 		}
 
 		return nil
 	}
 }
 
-func checkOutputWritable(outputPath string) error {
+// Ensures the path (or fallback) provided is writable. Returns path, error
+func ensureOutputWriteable(outputPath string, fallbackPath string) (string, error) {
+	// If no outputPath is provided use fallback path and track.
+	usingFallback := false
+	if outputPath == "" {
+		outputPath = fallbackPath
+		usingFallback = true
+	}
+
 	outputPath, err := homedir.Expand(outputPath)
 	if err != nil {
-		return err
+		return "", err
+	}
+
+	stat, err := os.Stat(outputPath)
+
+	// If the file doesn't exist, use the parent directory with given filename.
+	if os.IsNotExist(err) {
+		if err = unix.Access(path.Dir(outputPath), unix.W_OK); err != nil {
+			return "", fmt.Errorf("Output directory is not writable: %s", path.Dir(outputPath))
+		}
+		return outputPath, nil
+	} else if err != nil {
+		return "", fmt.Errorf("Unexpected error checking output path: %w", err)
 	}
 
-	// Check if the file exists
-	_, err = os.Stat(outputPath)
-	if err == nil {
-		// File exists, check if it's writable
-		return unix.Access(outputPath, unix.W_OK)
-	} else if os.IsNotExist(err) {
-		// File doesn't exist, check if the directory is writable
-		dir := filepath.Dir(outputPath)
-		return unix.Access(dir, unix.W_OK)
+	// If a directory was provided, use that with the fallback filename
+	if stat.IsDir() {
+		// If the fallback path already exists as a directory error.
+		if usingFallback {
+			return "", fmt.Errorf("Default output name \"%s\" conflicts with directory, provide --output", outputPath)
```

Comment:
No need to change, just a tip. But I recently discovered the format verb `%q` for wrapping the input with double quotes.